### PR TITLE
ref(api): parse pipeline tests, mql pipeline before treeify

### PR DESCRIPTION
**Pull request by kylemumma**
## Major Changes
* separate everything that comes before treeify in mql parse pipeline into its own stage
* add tests for this stage
* misc other things

### CODE REVIEW

File: snuba/query/dsl_mapper.py [hunk 1]
```diff
@@ -33,8 +31,7 @@
 
 def and_cond_repr(exp: Expression, visitor: ExpressionVisitor[str]) -> str:
     assert isinstance(exp, FunctionCall)
-    conditions = get_first_level_and_conditions(exp)
-    parameters = ", ".join([arg.accept(visitor) for arg in conditions])
+    parameters = ", ".join([arg.accept(visitor) for arg in exp.parameters])
```
Comment (kylemumma) [inline snuba/query/dsl_mapper.py:34]: the way mapper was creating conditions did not reflect the structure of the query object, bc it was rebalancing

Comment (kylemumma) [inline snuba/query/dsl_mapper.py:34]: the way mapper was creating conditions did not reflect the structure of the query object, bc it was rebalancing

File: snuba/utils/pipeline_logger.py
Comment (kylemumma) [inline snuba/utils/pipeline_logger.py:1]: this is a the class i used to capture intermediate values of the pipeline

File: tests/query/test_dsl_mapper.py
Comment (kylemumma) [inline tests/query/test_dsl_mapper.py:1]: this reflects the new behavior of the mapper to not rebalance

**Comment by volokluev** [comment]
cool title

File: snuba/query/mql/parser.py [hunk 2]
```diff
@@ -1124,3 +1132,37 @@ def parse_mql_query(
         _post_process(query, VALIDATORS)
 
     return query
+
+
+class MQLParseFirstHalf(
```
Comment (volokluev) [inline snuba/query/mql/parser.py:1137]: This name could be better. What does it actually do?

File: snuba/query/mql/parser.py [hunk 3]
```diff
@@ -1078,17 +1084,19 @@ def parse_mql_query(
...
-
-    if settings and settings.get_dry_run():
-        explain_meta.set_original_ast(str(query))
+    res = MQLParseFirstHalf().execute(
+        # query_settings and timer are dummy and dont matter
+        QueryPipelineResult(
+            data=(body, dataset, mql_context_dict, settings),
+            error=None,
+            query_settings=HTTPQuerySettings(),
+            timer=Timer("dummy"),
```
Comment (volokluev) [inline snuba/query/mql/parser.py:1093]:
```suggestion
            timer=Timer("mql_pipeline"),
```

File: snuba/utils/pipeline_logger.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        # the inner list is a "trace" (not sentry trace): -> t1 -> t2 -> t3 ->
+        # the tuple is (in, out)
+        self.logged: list[list[tuple[Any, Any]]] = []
+
+    def begin(self, val: Any) -> None:
+        """
+        Begins a new "trace" (not sentry trace): -> t1 -> t2 -> t3 ->
+        """
+        self.logged.append([])
+        self.isInput = True
```
Comment (volokluev) [inline snuba/utils/pipeline_logger.py:42]: what does this do?

**Review by volokluev — COMMENTED** [review]
Does the PipelineLogger have to be checked in? It's fine for an intermediate step but idk if I want it in the odebase as is

**Review by volokluev — APPROVED** [review]

**Pull request merged**